File: data-structure/python/lists/doubly-linked-list.py
class Node:
    def __init__ (self, value):
        self.value = value
        self.previous = None
        self.next = None

class DoublyLinkedList:
    def __init__ (self):
        self.head = None
        self.tail = None

    def addToHead(self, value):
        n = Node(value)

        if self.head == None:
            self.tail = n
        else:
            n.next = self.head
            self.head.previous = n
        self.head = n

    def addToTail(self, value):
        n = Node(value)

        if self.head == None:
            self.head = n
        else:
            n.previous = self.tail
            self.tail.next = n
        self.tail = n

    def showLtoR(self):
        if self.head == None:
            return None

        current = self.head
        while current != None:
            print(current.value)
            current = current.next

    def showRtoL(self):
        if self.head == None:
            return None

        current = self.tail
        while current != None:
            print(current.value)
            current = current.previous

    def removeFromHead(self):
        if self.head == None:
            print('The list is empty')
            return
        
        temp = self.head
        if self.head.next == None:
            self.tail = None
        else:
            temp.next.previous = None
            self.head = temp.next

        return temp
    
    def removeFromTail(self):
        if self.head == None:
            print('The list is empty')
            return
        
        temp = self.tail
        if self.tail.previous == None:
            self.head = None
        else:
            self.tail.previous.next = None

        self.tail = temp.previous
        return temp
    
    def remove(self, value):
        if self.head == None:
            print('The list is empty')
            return

        aux = self.head
        while aux and aux.value != value:
            aux = aux.next

        if aux is None:
            return None
        
        if aux == self.head:
            self.head = aux.next
            if self.head:
                self.head.previous = None

        if aux == self.tail:
            self.tail = aux.previous
            if self.tail:
                self.tail.next = None

        if aux.previous:
            aux.previous.next = aux.next
        if aux.next:
            aux.next.previous = aux.previous

        return aux.value
            


dll = DoublyLinkedList()
dll.addToHead(3)
dll.addToHead(2)
dll.addToHead(1)
dll.addToTail(4)
dll.remove(3)
dll.addToHead(0)
dll.addToTail(5)
dll.showLtoR()
print()
dll.showRtoL()

File: data-structure/js/lists/doubly-linked-list.ts
export class Node {
  public value: number;
  public next: Node | null = null;
  public previous: Node | null = null;
  constructor(value: number) {
    this.value = value;
  }
}

class DoublyLinkedList {
  private head: Node | null = null;
  private tail: Node | null = null;

  private isEmpty() {
    return this.head === null;
  }

  public unshift(value: number) {
    const node = new Node(value);

    if (this.isEmpty()) {
      this.tail = node;
    } else {
      this.head!.previous = node;
      node.next = this.head;
    }

    this.head = node;
  }

  public push(value: number) {
    const node = new Node(value);

    if (this.isEmpty()) {
      this.head = node;
    } else {
      node.previous = this.tail;
      this.tail!.next = node;
    }
    this.tail = node;
  }

  public readLToR() {
    if (!this.isEmpty()) {
      let current = this.head;

      while (current) {
        console.log(current.value);
        current = current.next!;
      }
    }
  }

  public readRToL() {
    if (!this.isEmpty()) {
      let current = this.tail;

      while (current) {
        console.log(current.value);
        current = current.previous!;
      }
    }
  }

  public shift() {
    if (this.isEmpty()) {
      console.log("The array is empty");
      return;
    }

    if (this.head === this.tail) {
      const value = this.head!.value;
      this.head = null;
      this.tail = null;
      return value;
    }

    if (this.head?.next) this.head.next.previous = null;

    const value = this.head?.value;
    this.head = this.head!.next;
    return value;
  }

  public pop() {
    if (this.isEmpty()) {
      console.log("The array is empty");
      return;
    }

    if (this.head === this.tail) {
      const value = this.head!.value;
      this.head = null;
      this.tail = null;
      return value;
    }

    if (this.tail!.previous) this.tail!.previous.next = null;

    const value = this.tail?.value;
    this.tail = this.tail!.previous;
    return value;
  }

  public remove(value: number) {
    if (this.isEmpty()) {
      console.log("The array is empty");
      return;
    }

    if (this.head === this.tail) {
      const value = this.head!.value;
      this.head = null;
      this.tail = null;
      return value;
    }

    let current = this.head;
    while (current && current.value !== value) {
      current = current.next;
    }

    if (!current) return null;

    if (current === this.head) {
      this.shift();
    } else if (current === this.tail) {
      this.pop();
    } else {
      current.next!.previous = current.previous;
      current.previous!.next = current.next;
    }

    return current.value;
  }
}

const dl = new DoublyLinkedList();
dl.unshift(3);
dl.unshift(2);
dl.unshift(1);
dl.push(4);
dl.push(5);
dl.unshift(0);
dl.pop();
dl.shift();
dl.remove(4);
dl.readLToR();
dl.readRToL();
